### DISCUSSION TITLE
trying openems with mpi. Not enough slots available error

### DISCUSSION THREAD

**Discussion by montanaviking** [thliebig/openEMS-Project#55] [Q&A Usage]
I am attempting to use openEMS mpi and I'm getting the following errors:
##############
Running remote openEMS_MPI in working dir: /tmp/openEMS_MPI_OxYoCbMoLCNW
warning: implicit conversion from numeric to char
warning: called from
    RunOpenEMS_MPI at line 90 column 15
    RunOpenEMS at line 82 column 9
    microstrip_mpi at line 174 column 1

Invalid MIT-MAGIC-COOKIE-1 keyInvalid MIT-MAGIC-COOKIE-1 keyInvalid MIT-MAGIC-COOKIE-1 keyInvalid MIT-MAGIC-COOKIE-1 key--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 4
slots that were requested by the application:

  /opt/openEMS/bin/openEMS

Either request fewer slots for your application, or make more slots
available for use.

A "slot" is the Open MPI term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which Open MPI processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, Open MPI defaults to the number of processor cores

In all the above cases, if you want Open MPI to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --oversubscribe option to ignore the
number of available slots when deciding the number of processes to
launch.
--------------------------------------------------------------------------
error: mpirun openEMS failed!
error: called from
    RunOpenEMS_MPI at line 97 column 5
    RunOpenEMS at line 82 column 9
    microstrip_mpi at line 174 column 1

###########################

My source is in Octave (Matlab) format and is shown below:

##############
%
% microstrip transmission line, Z is normal to substrate Y is the direction of propagation and X is the width
% try mpi

close all
clear
clc

% mpi setup

Settings.MPI.Binary = '/opt/openEMS/bin/openEMS';
Settings.MPI.NrProc = 4;
Settings.MPI.Hosts = {'wolfpack'};

........

%% run openEMS
%RunOpenEMS( Sim_Path, Sim_CSX, '--numThreads=4',Settings );
options='';
RunOpenEMS( Sim_Path, Sim_CSX,options,Settings );

...........

####################

Please note that this is running on machine 'hydra' and the remote 2nd machine is 'wolfpack'
Both hydra and wolfpack are 28-core machines but the above code works only when Settings.MPI.NrProc = 1;

I was not able to find the answer after an extensive search.
I'm stumped as to what I'm missing here and it's probably obvious to those with more experience than me.
Thanks in advance!
Phil

**Comment by 0xCoto** [comment]
(Never tried openEMS with MPI myself - I'm just subscribing to this thread. If you find a solution, it would be helpful if you could share it here as well. Very curious to hear what kind of speedup is achievable!)

**Reply by montanaviking** [reply]
Hi 0xCoto,
Thanks for responding. Also, my base program is running on hydra + wolfpack but I only specified wolfpack in the Settings.MPI.Hosts since the root process is running on hydra.
I'm hoping that adding extra machines will give a speed up since they are connected by 10G LANs and these used, two-socket XEON E5-26xx V4 or V3 servers are relatively cheap, while scaling to four sockets / machine adds a lot to cost. I suspect that more sockets = more speed?
If this works, and I get a good speed-up, I then have an excuse to load up on more machines. The nice thing is that I can power them on only when I need the extra compute power.
I'm suspicious that the problem isn't that the code thinks we're running out of cores, but rather something else that's give this as an error message?
Incidentally the error trace shows the initial error in the line 82 of RunOpenEMS_MPI.m where status is returning 0

#############################
% from RunOpenEMS_MPI.m
if isfield(Settings.MPI,'Hosts')
    disp(['Running remote openEMS_MPI in working dir: ' work_path]);
    [status]  = system(['mpiexec ' Settings.MPI.GlobalArgs ' -host ' HostList ' -n ' int2str(NrProc) ' -wdir ' work_path ' ' Settings.MPI.Binary ' ' Sim_File ' ' opts ' ' append_unix]);
else
    disp('Running local openEMS_MPI');
    [status]  = system(['mpiexec ' Settings.MPI.GlobalArgs ' -n ' int2str(NrProc) ' ' Settings.MPI.Binary ' ' Sim_File ' ' opts ' ' append_unix]);
end

if (status~=0)
    error('openEMS:RunOpenEMS','mpirun openEMS failed!');
end
############################################################

Thanks,
Phil
